app/models/donation.rb

```diff
@@ -0,0 +1,92 @@
+class Donation
+  include ActiveModel::Model
+
+  attr_accessor :email, :employer, :occupation, :stripe_token, :recurring,
+    :utm_source, :utm_medium, :utm_campaign, :source_url, :amount_in_cents
+
+  validates :email, presence: true,
+    email_format: { message: 'is not a valid email' }
+  validates :employer, presence: true
+  validates :occupation, presence: true
+  validates :stripe_token, presence: true
+  validates :amount_in_cents, presence: true,
+    numericality: { greater_than: 0, only_integer: true }
+
+  def process!
+    if valid?
+      process_payment
+      record_donation
+      create_donate_action
+    else
+      false
+    end
+
+  rescue Stripe::CardError => e
+    errors.add(:stripe_token, e.json_body[:error][:message])
+    false
+  end
+
+  private
+
+  def process_payment
+    if recurring
+      stripe_customer = StripeCustomer.new(create_stripe_customer)
+      person.update(stripe_id: stripe_customer.id)
+      person.create_subscription(remote_id: stripe_customer.subscription_id)
+    else
+      create_stripe_charge
+    end
+  end
+
+  def record_donation
+    person = { email: email, employer: employer, occupation: occupation }
+    NbDonationCreateJob.perform_later(amount_as_integer, person)
+  end
+
+  def create_donate_action
+    person.create_action(utm_source: utm_source,
+                         utm_medium: utm_medium,
+                         utm_campaign: utm_campaign,
+                         source_url: source_url,
+                         template_id: 'donate',
+                         donation_amount_in_cents: amount_as_integer)
+  end
+
+  def person
+    @person ||= find_or_create_person
+  end
+
+  def find_or_create_person
+    person = Person.find_or_initialize_by(email: email)
+    person.update(skip_nb_update: true)
+    person
+  end
+
+  def create_stripe_customer
+    Stripe::Customer.create(source: stripe_token,
+                            plan: 'one_dollar_monthly',
+                            email: email,
+                            quantity: amount_as_integer/100)
+  end
+
+  def create_stripe_charge
+    Stripe::Charge.create(amount: amount_as_integer,
+                          source: stripe_token,
+                          currency: 'usd',
+                          description: "donation from #{email}")
+  end
+
+  def amount_as_integer
+    amount_in_cents.to_i
+  end
+
+  StripeCustomer = Struct.new(:stripe_customer) do
```

Comment:
There's no compelling reason to go with struct instead of class, except that struct is simpler than class and I was following an example that used a struct. I got the idea from Sandi Metz's POODR book. I started with a struct because that's the simplest thing that will do the job, thinking I would create a new class (in separate file) when I have a better idea of what's needed. I can change it to a class now if you find that more readable.

In the short term it's a lot of extra work for not much gain, but the idea is that if we don't do it now our objects will become entangled (which I think is the case with our existing code). The end goal is to isolate interfaces that might change by writing new methods or adapter classes and to limit what classes need to know about the internals of other classes. The goal in this case is to have a layer of abstraction between our models and Stripe such that we can later switch to a new payment processor and not have to change all our code.